### ISSUE TITLE
Tax information is not being passed through to PayPal

### ISSUE THREAD

**Issue by ValeDeOro** [woocommerce/woocommerce#17969]
When using PayPal Standard as the Payment method and using tax settings, the tax information is not passed to PayPal. 
The right amount is charged, BUT the PayPal transaction details ignore that part of the amount is supposed to be tax. This can cause problems for accounting purposes. 

Example: 
This order includes tax: 
![]( http://cld.wthms.co/WLoEOT+)
Link to screenshot:  http://cld.wthms.co/WLoEOT

This is the information in Paypal: 
![](http://cld.wthms.co/hhAhZ6+)
Link to screenshot: http://cld.wthms.co/hhAhZ6

The final amount is correct, but the Payment Details section in PayPal does not mirror the information of the WooCommerce order. 

@gedex confirmed that this is a problem coming from the way WooCommerce Standard passes the data to PayPal, hence logging it here.

**Comment by claudiulodro** [comment]
I can see it going through properly in the PayPal sandbox when I try it on my local environment:

Screenshot from PayPal:
<img width="313" alt="screen shot 2017-12-01 at 12 20 58 pm" src="https://user-images.githubusercontent.com/7317227/33501738-83cb8ccc-d692-11e7-89a0-b04a323676c1.png">

And you can see the taxes in the generated URL for the PayPal Standard checkout:
`https://www.sandbox.paypal.com/cgi-bin/webscr?test_ipn=1&cmd=_cart&business={redacted}&no_note=1&currency_code=USD&charset=utf-8&rm=1&upload=1&return=http%3A%2F%2Flocal.wordpress.dev%2Fcheckout%2Forder-received%2F256%3Fkey%3Dwc_order_5a21b7a41fe6f%26utm_nooverride%3D1&cancel_return=http%3A%2F%2Flocal.wordpress.dev%2Fcart%2F%3Fcancel_order%3Dtrue%26order%3Dwc_order_5a21b7a41fe6f%26order_id%3D256%26redirect%26_wpnonce%3Df5e2f044d8&page_style=&image_url=&paymentaction=sale&bn=WooThemes_Cart&invoice=WC-256&custom=%7B%22order_id%22%3A256%2C%22order_key%22%3A%22wc_order_5a21b7a41fe6f%22%7D&notify_url=http%3A%2F%2Flocal.wordpress.dev%2Fwc-api%2FWC_Gateway_Paypal%2F&first_name=claud&last_name=lodro&address1=123+Test+St.&address2=&city=Portland&state=OR&zip=97266&country=US&email=test%40example.com&night_phone_a=503&night_phone_b=123&night_phone_c=4567&no_shipping=1&`**tax_cart=3.00**`&item_name_1=Album&quantity_1=1&amount_1=15&item_number_1=woo-album`

A quick Google search for that product took me to [this site](https://mmbhof.org/product/signed-fat-tire-flyer-repack-and-the-birth-of-mountain-biking/) which had a ton of errors when trying to check out (the errors only happen in new sessions or incognito mode). Whatever is causing those errors is most likely causing the tax issues in your screenshot. I wasn't able to get a proper tax calculation to show up while trying out the Checkout on that site.

<img width="825" alt="screen shot 2017-12-01 at 12 28 24 pm" src="https://user-images.githubusercontent.com/7317227/33502035-7bf1a1ca-d693-11e7-8f4d-db208d59108c.png">

@gedex were you able to recreate the problem or did you notice any part in the codebase that would cause the problem? Everything seems to be working great as far as I can tell.

**Comment by Farallon5** [comment]
The errors referenced above have been corrected, and subsequent test transactions have shown the tax problem remains exactly as before. ie: Tax shows up properly in the Woo Cart, but Paypal does not show the tax after the transaction is finalized.

**Comment by gedex** [comment]
> @gedex were you able to recreate the problem or did you notice any part in the codebase that would cause the problem? Everything seems to be working great as far as I can tell.

@claudiulodro So when creating line items request to PayPal, we did this check:

https://github.com/woocommerce/woocommerce/blob/bf6c39fbf2c1362171d49ff33a8ca2679bf8f7ec/includes/gateways/paypal/includes/class-wc-gateway-paypal-request.php#L173

if the `if` condition evaluates to false, it will group all items into one and shipping amount if shipping is enabled but no tax info being set in the arg. So, to reproduce that is to fail the `if` condition. I'm using tax setting to fail `( ! wc_tax_enabled() || ! wc_prices_include_tax() )`. Here's my tax setting:

<img width="663" alt="screen shot 2017-12-03 at 12 45 58 am" src="https://user-images.githubusercontent.com/78313/33518173-5adad084-d7c3-11e7-91c7-36ac8027397a.png">

Here's the order looks like:

<img width="331" alt="screen shot 2017-12-03 at 12 51 28 am" src="https://user-images.githubusercontent.com/78313/33518236-3e07d5c8-d7c4-11e7-9bbd-97514bcb2760.png">

With that, I got:

<img width="425" alt="screen shot 2017-12-03 at 12 52 05 am" src="https://user-images.githubusercontent.com/78313/33518238-445cb696-d7c4-11e7-8bb7-4fee524395a9.png">

**Comment by Farallon5** [comment]
It looks like you have confirmed there is a problem, right?

**Comment by mikejolley** [comment]
I've looked into this again (did back in 2012 too https://github.com/woocommerce/woocommerce/issues/951) and there are reasons it's not broken down and sent.

- When you send the order total as a whole (likely to avoid rounding issues), breaking out the tax from the shipping and total can introduce more rounding issues so payment will fail.
- You cannot send shipping tax separately to item tax, so in most cases you get a confusing display of _Item Total, followed by VAT, followed by P&P_ (it is not clear where the tax comes from). In WC we show shipping before taxes.

In the PayPal payment page, omitting tax (which is optional) is fine because the row just gets hidden. This only seems to be a problem on the report pages within PayPal so only visible by the account owner. Perhaps PayPal could look into this and adjust (hide 0 rows, or omit if not provided?). I'll send this as a request to Ekta at PayPal to get some feedback, but we won't be able to adjust this in core yet.

FWIW if this has only recently become a problem, it may be a different rounding issue trigger the 'send all items as a single total' logic. Possibly related to https://github.com/woocommerce/woocommerce/issues/17970

Thanks

**Comment by Farallon5** [comment]
I don't get it. Isn't this (or couldn't it be) just a simple 
field-mapping process to make sure the source woo fields get mapped to 
the destination paypal fields? I don't understand why there would be 
rounding issues if the amounts in each field were just sent as-is.  We 
are not talking about very many fields per transaction.

Marc


------ Original Message ------

**Comment by mikejolley** [comment]
The rounding issues come into play because PayPal API does not support prices including tax.

**Comment by Farallon5** [comment]
We want to keep all the prices sent to paypal to be excluding tax. We want paypal to show the tax as a separate item, just like the woo invoice #9086 above.

**Comment by mikejolley** [comment]
And that depends on 2 things; 

1, if a rounding error is detected and 
2, if your product prices are defined including or excluding tax. 

If the problem is 1, thats a separate issue with the order calculation logic and one thats being worked on elsewhere.

**Comment by mikaey** [comment]
It seems like there's two situations that Woo has to deal with: when item prices are inclusive of tax, and when item prices are exclusive of tax.  Both are tricky to handle.

PayPal has two ways to provide tax amounts: one is to just give us the grand total of taxes on the order; the other is to give us the tax amounts for each line item on the order.  The second method is more complicated, because you *still* have to provide the grand total of taxes on the order, and it has to match up with the sum of the line item taxes.  On top of that, you have to tell us what the tax is for an individual quantity of that item.  This becomes a problem if the buyer purchases more than one quantity of an item: for example, if a buyer purchases 500 items at $1.37 each, and a 7% tax applies to the order, then the tax per item would be $0.0959; but since PayPal can't accept more than two decimal places, you would have to tell PayPal that the tax per quantity is either $0.09 or $0.10.  But that's not right -- because the grand total of tax for a quantity of 500 is $47.95; using $0.09 per quantity would yield $45, while using $0.10 per quantity would yield $50.  Since WooCommerce can't reliably provide item-level tax this way every time, it's safer to just supply the order-level tax instead.  (On top of that, we don't show the item-level tax to the buyer, so there isn't really much lost).

So, for situations where the item prices are exclusive of tax, then passing in the tax is pretty trivial.  You'll need to calculate the tax grand total (e.g., item tax + shipping tax + any other tax that may be applicable) and pass it in `tax_cart`, as @claudiulodro noted above.

When item prices are inclusive of tax is when things get tricky.  PayPal operates off the idea that item prices are going to be exclusive of tax.  There's no way to say "here's how much tax I'm charging, but just show it to the user -- don't use it when calculating totals" -- if tax is provided to us, we're going to add it in to the grand total.  This leaves you with a couple of choices: pass the item prices to us as they are in WooCommerce (inclusive of tax), or separate the tax out and pass us the price (exclusive of tax) and the grand total of tax separately.  I'm not even sure that WooCommerce knows what the item price would be exclusive of tax, so that leaves you with just the first option.  The downside here is that the tax will show on the receipt as $0.00 every time.

To summarize, I think my advice would be as follows:

- If item prices are exclusive of tax, then calculate a grand total of all taxes and pass that to us in `tax_cart`.
- If item prices are inclusive of tax, then just add the item prices together and don't worry about passing tax to us.

**Comment by Farallon5** [comment]
"The downside here is that the tax will show on the receipt as $0.00 every time." Yep, that is a huge downside and would be unacceptable from a customer service standpoint.

**Comment by mikejolley** [comment]
@mikaey The other issue with `tax_cart` is you cannot also pass in tax_shipping, so you have to lump it. And when you lump it, it displays everything out of order. It lists items, then VAT, then shipping.
